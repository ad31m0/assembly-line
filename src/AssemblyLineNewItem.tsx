import { useState } from 'react';
import { moveItemType } from "./utils/moveItem";

export const AssemblyLineNewItem = ({ moveItem }: { moveItem: moveItemType }) => {
  const [newItemName, setNewItemName] = useState("");
  return (<div>
    <span>Add an item: </span>
    <input
      maxLength={160}
      value={newItemName}
      onChange={(event) => {
        setNewItemName(event.target.value);
      }}
      onKeyPress={(event) => {
        if (event.code === "Enter" && newItemName !== "") {
          moveItem(-1, -1, +1, newItemName);
          setNewItemName("");
        }
      }}
    />
  </div>)
}  