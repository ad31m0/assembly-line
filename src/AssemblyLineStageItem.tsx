import { moveItemType } from "./utils/moveItem";

export const AssemblyLineStageItem = ({ itemIdx, itemName, stageIdx, moveItem }: { itemIdx: number, itemName: string, stageIdx: number, moveItem: moveItemType }) => {
  return (
    <div key={`${stageIdx}_${itemIdx}`}>
      <button
        onMouseDown={(event) => {
          switch (event.button) {
            case 0:
              moveItem(itemIdx, stageIdx, +1);
              break;
            case 2:
              moveItem(itemIdx, stageIdx, -1);
              break;
            default:
              // unsupported do nothing;
              break;
          }
        }}>
        {itemName}
      </button>
    </div>
  );
}