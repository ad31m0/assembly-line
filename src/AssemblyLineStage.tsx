import { AssemblyLineStageItem } from "./AssemblyLineStageItem"
import { moveItemType } from "./utils/moveItem"

export const AssemblyLineStage = ({ stageIdx, stageName, stageItems, moveItem }: { stageIdx: number, stageName: string, stageItems: string[], moveItem: moveItemType }) => {
  return (<div className="assembly-line-stage" style={{ width: "200px" }}>
    <h4>{stageName}</h4>
    {stageItems.map((itemName: string, itemIdx: number) =>
      <AssemblyLineStageItem {...{ key: `stage-${stageIdx}-item-${itemIdx}`, itemIdx, itemName, stageIdx, moveItem }} />
    )}
  </div>)
}