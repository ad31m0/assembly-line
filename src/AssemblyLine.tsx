import { AssemblyLineNewItem } from "./AssemblyLineNewItem";
import { moveItemCtor } from "./utils/moveItem";
import { disableContextMenu } from './utils/disableContextMenu';
import { AssemblyLineStage } from './AssemblyLineStage';
import { useLocalStorage } from "./utils/useLocalStorage";

export const AssemblyLine = ({ stages, items }: { stages: string[], items?: string[][] }) => {
  items = items ?? stages.map((name: string, idx: number) => []);
  const [stagesItems, setStagesItems] = useLocalStorage("items", items);
  const moveItem = moveItemCtor({ stagesItems, setStagesItems });

  return (
    <div className="assembly-line" onContextMenu={disableContextMenu}>
      <AssemblyLineNewItem {...{ moveItem }} />
      <br/>
      <hr/>
      <div className="assembly-line-pipeline hs">
        {stagesItems.map((stageItems: string[], stageIdx: number) =>
          <AssemblyLineStage {...{ key: `stage-${stageIdx}`, stageIdx, stageName: stages[stageIdx], stageItems, moveItem }} />
        )}
      </div>
    </div>
  );
}