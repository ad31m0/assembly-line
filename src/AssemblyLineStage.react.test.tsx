import React from 'react';
// setup file
import { configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });
// test file
import { shallow, mount, render } from 'enzyme';
import { AssemblyLine } from './AssemblyLine';
import { AssemblyLineStageItem } from './AssemblyLineStageItem';
import { AssemblyLineNewItem } from './AssemblyLineNewItem';
import { AssemblyLineStage } from './AssemblyLineStage';


describe('<AssemblyLine /> with no props', () => {
    const container = shallow(<AssemblyLineStage  {...{ stageIdx: 0, stageName: "Idea", moveItem: () => { }, stageItems: ["1"] }} />);

    it('should match the snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });
    it('should have a add new item input', () => {
        expect(container.find(AssemblyLineStageItem).length).toEqual(1);
    });

});
