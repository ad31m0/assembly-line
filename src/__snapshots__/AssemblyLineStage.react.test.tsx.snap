// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssemblyLine /> with no props should match the snapshot 1`] = `"<div class=\\"assembly-line-stage\\" style=\\"width:200px\\"><h4>Idea</h4><div><button>1</button></div></div>"`;
